### PR TITLE
Remove native_functions.yaml dependency from IndexKernel.{cpp,cu}

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67875
* #67874
* #67721
* #66980
* #66979
* #67216
* __->__ #66914
* #66913

Differential Revision: [D31856105](https://our.internmc.facebook.com/intern/diff/D31856105)

### CODE REVIEW

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Review by dagitses — APPROVED** [review]

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

File: aten/src/ATen/native/cpu/IndexKernel.cpp [hunk 1]
```diff
@@ -116,9 +118,9 @@ struct IndexToOffset {
   const IntArrayRef sizes;
   const IntArrayRef strides;
   const int ndim;
-  IndexToOffset(const Tensor & tensor) : sizes(tensor.sizes()),
-                                         strides(tensor.strides()),
-                                         ndim(tensor.dim()) {}
+  IndexToOffset(const TensorBase & tensor) : sizes(tensor.sizes()),
```
Comment (dagitses) [inline aten/src/ATen/native/cpu/IndexKernel.cpp:121]: internal lint is complaining about this needing to be marked explicit. I know it's not your code but might as well make it happy.

File: aten/src/ATen/native/cpu/IndexKernel.cpp [hunk 2]
```diff
@@ -116,9 +118,9 @@ struct IndexToOffset {
...
+                                             strides(tensor.strides()),
+                                             ndim(tensor.dim()) {}
```
Comment (dagitses) [inline aten/src/ATen/native/cpu/IndexKernel.cpp:123]: internal lint is complaining about narrowing from 64-bit to 32-bit here. Make ndim member 64-bit?

File: aten/src/ATen/native/cuda/IndexKernel.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    : c10::MaybeOwned<Tensor>::borrowed(mask);
+  auto self_temp = (self.dim() == 0)
+    ? c10::MaybeOwned<Tensor>::owned(self.unsqueeze(0))
+    : c10::MaybeOwned<Tensor>::borrowed(self);
+
+  // Cannot reassign to mask_temp and self_temp here! if they are
+  // owning and expand_outplace returns a borrow, the returned borrow
+  // would dangle.
+  auto mask_self_expanded = expand_outplace(*mask_temp, *self_temp);
+  at::native::index_out(result, *std::get<1>(mask_self_expanded), c10::List<c10::optional<at::Tensor>>({*std::get<0>(std::move(mask_self_expanded))}));
```
Comment (dagitses) [inline aten/src/ATen/native/cuda/IndexKernel.cpp:31]: from internal lint:
> Potential variable access after move.
> Variable mask_self_expanded is accessed here, and is moved-from in an unspecified order

**Review by dagitses — APPROVED** [review]

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

File: aten/src/ATen/native/cuda/IndexKernel.cpp (hunk 3 above)
Comment (dagitses) [reply aten/src/ATen/native/cuda/IndexKernel.cpp]: i see a formatting improvement, but I don't see any material changes here.

Comment (peterbell10) [reply aten/src/ATen/native/cuda/IndexKernel.cpp]: The `std::move` now comes after the `std::get`, so only one tuple element is created as an r-value reference.

Comment (dagitses) [reply aten/src/ATen/native/cuda/IndexKernel.cpp]: ah thanks, i see that now.

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Comment by dagitses** [comment]
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856105).

**Pull request closed without merging**